ipaserver/install/replication.py

```diff
@@ -476,61 +476,69 @@ def set_replica_binddngroup(self, r_conn, entry):
-            pass
-        else:
-            binddns = entry.setdefault('nsDS5ReplicaBindDN', [])
-            if replica_binddn not in {DN(m) for m in binddns}:
-                # Add the new replication manager
-                binddns.append(replica_binddn)
-            for key, value in REPLICA_CREATION_SETTINGS.items():
-                entry[key] = value
+            # no entry, create new one
+            entry = conn.make_entry(
+                dn,
+                objectclass=["top", "nsds5replica", "extensibleobject"],
+                cn=["replica"],
+                nsds5replicaroot=[str(self.db_suffix)],
+                nsds5replicaid=[str(replica_id)],
+                nsds5replicatype=[self.get_replica_type()],
+                nsds5flags=["1"],
+                nsds5replicabinddn=[replica_binddn],
+                nsds5replicabinddngroup=[self.repl_man_group_dn],
+                nsds5replicalegacyconsumer=["off"],
+                **REPLICA_CREATION_SETTINGS
+            )
             try:
-                conn.update_entry(entry)
-            except errors.EmptyModlist:
-                pass
+                conn.add_entry(entry)
+            except errors.DuplicateEntry:
+                logger.debug("Lost race against another replica, updating")
+                # fetch entry that have been added by another replica
+                entry = conn.get_entry(dn)
+            else:
+                logger.debug("Added replica config %s", dn)
+                # added entry successfully
+                return entry
 
-            self.set_replica_binddngroup(conn, entry)
+        # either existing entry or lost race
+        binddns = entry.setdefault('nsDS5ReplicaBindDN', [])
+        if replica_binddn not in {DN(m) for m in binddns}:
+            # Add the new replication manager
+            binddns.append(replica_binddn)
 
-            # replication is already configured
-            return
+        for key, value in REPLICA_CREATION_SETTINGS.items():
+            entry[key] = value
 
-        replica_type = self.get_replica_type()
+        try:
+            conn.update_entry(entry)
+        except errors.EmptyModlist:
+            logger.debug("No update to %s necessary", entry.dn)
+        else:
+            logger.debug("Update replica config %s", entry.dn)
 
-        entry = conn.make_entry(
-            dn,
-            objectclass=["top", "nsds5replica", "extensibleobject"],
-            cn=["replica"],
-            nsds5replicaroot=[str(self.db_suffix)],
-            nsds5replicaid=[str(replica_id)],
-            nsds5replicatype=[replica_type],
-            nsds5flags=["1"],
-            nsds5replicabinddn=[replica_binddn],
-            nsds5replicabinddngroup=[self.repl_man_group_dn],
-            nsds5replicalegacyconsumer=["off"],
-            **REPLICA_CREATION_SETTINGS
-        )
-        conn.add_entry(entry)
+        self._set_replica_binddngroup(conn, entry)
+
+        return entry
 
```

Comment:
This part of the code looks good to me but I think it will fail at the next step: setup_changelog.

If two ipa-replica-install are doing the setup_promote_replication at the same time. The patch prevents a failure at replica_config (catching the replica already existing). So one did a ADD and the other will check the binddngroup. Next step is setup_changelog. We should also harden setup_changelog from running in parallel.

Reply:
``setup_changelog`` already handles add conflicts and catches the duplicate entry error.

https://github.com/freeipa/freeipa/blob/b21c3f9e88ea0b54140be02ba72c7076f05c8209/ipaserver/install/replication.py#L543-L563

Reply:
Right. One place I found fragile is ensure_replication_managers when it calls _add_replication_managers. The replication managers entry can already exist

https://github.com/freeipa/freeipa/blob/b21c3f9e88ea0b54140be02ba72c7076f05c8209/ipaserver/install/replication.py#L1671-1693

Reply:
Memo to me: Look into ``ldap.TYPE_OR_VALUE_EXISTS``